api/envoy/api/v2/cds.proto

```diff
@@ -632,6 +641,33 @@ message Cluster {
   // If this flag is not set to true, Envoy will wait until the hosts fail active health
   // checking before removing it from the cluster.
   bool drain_connections_on_host_removal = 32;
+
+  // New mechanism for LB policy configuration. Used only if the
+  // :ref:`lb_policy<envoy_api_field_Cluster.lb_policy>` field has the value
+  // :ref:`NEW_POLICY_CONFIG<envoy_api_enum_value_Cluster.LbPolicy.NEW_POLICY_CONFIG>`.
+  LoadBalancingPolicy load_balancing_policy = 40;
+}
+
+// New Load balancing policy selection.
+//
+// Every LB policy defined via this mechanism will be identified via a
+// unique name using reverse DNS notation. If the policy needs configuration
+// parameters, it must define a message for its own configuration, which will
+// be stored in the config field. The name of the policy will tell clients
+// which type of message they should expect to see in the config field.
+message LoadBalancingPolicy {
+  message Policy {
+    // Required. The name of the LB policy.
+    string name = 1;
+    // Optional config for the LB policy.
+    // No more than one of these two fields may be populated.
+    google.protobuf.Struct config = 2;
+    google.protobuf.Any typed_config = 3;
+  }
+  // Each client will iterate over the list in order and stop at the first
+  // policy that it supports. This provides a mechanism for starting to use
+  // new LB policies that are not yet supported by all clients.
```

Comment:
This seems very strange. Its almost saying here is an API way of fixing backward compatibility issues that any control plane should take care of. IOW, you are pushing what should be the control plane's responsibility in detecting the envoy version and shipping the right config down into the data plane. There is nothing wrong with that, except it has to be consistent throughout the API. Everywhere else in the API, we strive for backward compatibility, not forward compatibility. If we take the approach of reducing control plane overhead, then we should start apply it equally and not just in LB selection.

Reply:
As a general API principle, we want to move towards a model where the control plane has to do relatively little per-client processing, so to the extent we can specify sensible fallbacks, this works. It certainly doesn't work everywhere, but for LB policy it's sensible. I agree with your concern over consistency; directionally we want to move here and will aim to do that in future major versions.

Reply:
but its not that simple. Imagine control plane shipping a 2x set of listeners (set 1 with old field formats/algorithm choices, etc., set 2 with newer ones), where new ones have backwards incompatible fields. Per your (new) principle, older Envoy should ignore those erroneous listeners and only take in listeners that actually work, because this older Envoy has "forward compatibility". Now, the newer Envoy also has backwards compatibility. So, this newer Envoy will be able to process both the older set and the newer set, both of which have potentially the same settings (like binding to same addresses, or what not). So, which version is the newer envoy going to pick? Old or new or merge both ?

this is a construed scenario but you get the idea. This forward compatibility and backward compatibility has far reaching implications than just the LB. If we cannot allow forward compatibility (other than simply ignoring unknown fields), then we shouldn't be allowing it here either. If we can allow forward compatibility across the board, then we should figure out the backward compatibility issue before taking this big step.  Else, this will become one of those futuristic api additions that never reaches reality, and forever confuses users as to the utility of the field.

Reply:
Put another way, to sovle this forward+backward compat issue, the control plane will have to know the version of the client to ship it the only latest possible configuration. At that point, the goal of this complexity introduced in the data plane API becomes moot as the primary target user (the control plane) cannot use it in reliable fashion.

Reply:
I don't understand the point about backwards incompatible fields. We're planning on ending completely backwards incompatible changes within a major version, see https://github.com/envoyproxy/envoy/issues/6271, this is what we're doing in v3.

Reply:
@rshriram are you OK with us merging this as is for v2?